File: src/main/java/org/jesperancinha/airports/containers/MainContainerService.java
package org.jesperancinha.airports.containers;

import lombok.Getter;
import lombok.Setter;
import org.jesperancinha.airports.pojos.Airport;
import org.jesperancinha.airports.pojos.Country;
import org.jesperancinha.airports.pojos.Runway;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by joaofilipesabinoesperancinha on 08-08-16.
 */
@Getter
@Setter
@Service
public class MainContainerService {

    private AirportContainer airportContainer = new AirportContainer();

    private CountryContainer countryContainer = new CountryContainer();

    private RunwayContainer runwayContainer = new RunwayContainer();

    private List<Airport> fullAiportInfo;

    public MainContainerService() throws IOException {
        List<Airport> airports = airportContainer.getAirports();
        List<Country> countries = countryContainer.getCountries();
        List<Runway> runways = runwayContainer.getRunways();

        airports.sort(Comparator.comparing(Airport::getIsoCountry));
        countries.sort(Comparator.comparing(Country::getCode));

        fullAiportInfo = airports
                .parallelStream()
                .peek(airport -> airport
                        .setCountry(countries.stream()
                                .filter(country -> country.getCode().equals(airport.getIsoCountry()))
                                .findFirst()
                                .orElse(null)))
                .collect(Collectors.toList());

        fullAiportInfo = airports
                .parallelStream()
                .peek(airport -> airport
                        .addRunWay(runways.stream()
                                .filter(runway -> runway.getAirportIdent().equals(airport.getIdent()))
                                .findFirst()
                                .orElse(null)))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/org/jesperancinha/airports/containers/MainContainerService.kt
package org.jesperancinha.airports.containers

import org.jesperancinha.airports.pojos.Airport
import org.jesperancinha.airports.pojos.Country
import org.jesperancinha.airports.pojos.Runway
import org.springframework.stereotype.Service
import java.util.stream.Collectors

/**
 * Created by joaofilipesabinoesperancinha on 08-08-16.
 */
@Service
class MainContainerService(
    var fullAiportInfo: List<Airport>
){
    private val airportContainer: AirportContainer = AirportContainer()
    private val countryContainer: CountryContainer = CountryContainer()
    private val runwayContainer: RunwayContainer = RunwayContainer()
    init {
        val airports = airportContainer.airports
        val countries = countryContainer.countries
        val runways = runwayContainer.runways
        fullAiportInfo = getFullAirportInfo(airports, countries, runways)

    }

    private fun getFullAirportInfo(
        airports: List<Airport>,
        countries: List<Country>,
        runways: List<Runway>
    ) = airports
        .sortedBy { it.isoCountry }
        .map { airport: Airport ->
            airport.country = countries
                .sortedBy { it.code }
                .stream()
                .filter { country -> country.code == airport.isoCountry }
                .findFirst()
                .orElse(null)
            airport
        }
        .map { airport: Airport ->
            val extraRunway = runways.firstOrNull { runway -> runway.airportIdent == airport.ident }
            extraRunway?.let {
                airport
                    .addRunWay(extraRunway)
            }
            airport
        }
}